### PR TITLE
BUG: to_datetime with M or Y unit and non-round float

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [ ] closes #xxxx (Replace xxxx with the GitHub issue number)
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

**Comment by MarcoGorelli** [comment]
If non-round values aren't allowed, would this supersede https://github.com/pandas-dev/pandas/pull/50183?

**Comment by jbrockmendel** [comment]
> If non-round values aren't allowed, would this supersede https://github.com/pandas-dev/pandas/pull/50183?

Its only for unit="M" or unit="Y"

File: pandas/tests/tools/test_to_datetime.py [hunk 1]
```diff
@@ -1421,6 +1421,28 @@ def test_to_datetime_fixed_offset(self):
 
 
 class TestToDatetimeUnit:
+    @pytest.mark.parametrize("unit", ["Y", "M"])
+    def test_to_datetime_month_or_year_unit_non_round_float(self, cache, unit):
```
Comment (MarcoGorelli) [inline pandas/tests/tools/test_to_datetime.py:1425]: gh number, I guess?

File: pandas/tests/tools/test_to_datetime.py [hunk 2]
```diff
@@ -1421,6 +1421,28 @@ def test_to_datetime_fixed_offset(self):
...
+        with pytest.raises(ValueError, match=msg):
+            to_datetime([1.5], unit=unit, errors="ignore")
+
+        res = to_datetime([1.5], unit=unit, errors="coerce")
+        expected = Index([NaT], dtype="M8[ns]")
+        tm.assert_index_equal(res, expected)
+
+        # round floats are OK
+        res = to_datetime([1.0], unit=unit)
+        expected = to_datetime([1], unit=unit)
```
Comment (MarcoGorelli) [inline pandas/tests/tools/test_to_datetime.py:1443]: shall we construct this explicitly as
```
expected = DatetimeIndex(['1970-01-31 10:29:06'], dtype='datetime64[ns]', freq=None)
```
?

File: pandas/_libs/tslib.pyx [hunk 3]
```diff
@@ -345,6 +349,18 @@ def array_with_unit_to_datetime(
                 if val != val or val == NPY_NAT:
                     iresult[i] = NPY_NAT
                 else:
+                    if is_ym and is_float_object(val) and val != int(val):
```
Comment (MarcoGorelli) [inline pandas/_libs/tslib.pyx:352]: nothing major, but I think `val % 1` would be more performant (and IMO clearer)
```python
In [9]: %%timeit
   ...: if val != int(val): pass
   ...: 
   ...: 
67.4 ns ± 4.53 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)

In [10]: %%timeit
    ...: if val % 1: pass
    ...: 
    ...: 
34.8 ns ± 1.49 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)
```

File: pandas/_libs/tslib.pyx [hunk 4]
```diff
@@ -412,7 +454,7 @@ def array_with_unit_to_datetime(
                 oresult[i] = <object>NaT
             else:
                 try:
-                    oresult[i] = Timestamp(cast_from_unit(val, unit))
+                    oresult[i] = Timestamp(val, unit=unit)
                 except OverflowError:
                     oresult[i] = val
```
Comment (MarcoGorelli) [inline pandas/_libs/tslib.pyx:459]: would it work to just catch `ValueError` here too, to address
```
to_datetime([1.5], unit='M', errors="ignore")
```
raising?

File: pandas/_libs/tslib.pyx [hunk 5]
```diff
@@ -361,8 +377,33 @@ def array_with_unit_to_datetime(
...
+                        if is_raise:
+                            raise ValueError(
+                                f"non convertible value {val} with the unit '{unit}'"
+                            )
+                        elif is_ignore:
+                            raise AssertionError
+                        iresult[i] = NPY_NAT
+                        continue
+
+                    if is_ym and fval != int(fval):
```
Comment (MarcoGorelli) [inline pandas/_libs/tslib.pyx:393]: same comment as above

**Review by MarcoGorelli — CHANGES_REQUESTED** [review]

File: pandas/_libs/tslib.pyx (hunk 3 above)
Comment (jbrockmendel) [reply pandas/_libs/tslib.pyx]: updated to use `val.is_integer()`, maximally clear, _almost_ as performant as mod.

File: pandas/_libs/tslib.pyx (hunk 4 above)
Comment (jbrockmendel) [reply pandas/_libs/tslib.pyx:459]: id like to punt on this for the time being.   i think this path will need a revamp to be more consistent with non-unit path anyway

Comment (MarcoGorelli) [reply pandas/_libs/tslib.pyx:459]: sure thing

File: pandas/_libs/tslib.pyx (hunk 3 above)
Comment (MarcoGorelli) [reply pandas/_libs/tslib.pyx]: nice!

File: doc/source/whatsnew/v2.0.0.rst [hunk 6]
```diff
@@ -489,6 +489,7 @@ Other API changes
   new DataFrame (shallow copy) instead of the original DataFrame, consistent with other
   methods to get a full slice (for example ``df.loc[:]`` or ``df[:]``) (:issue:`49469`)
 - Disallow computing ``cumprod`` for :class:`Timedelta` object; previously this returned incorrect values (:issue:`50246`)
+- :func:`to_datetime` with ``unit`` of either "Y" or "M" will now raise if a sequence contains a non-round ``float`` value, matching the ``Timestamp`` behavior (:issue:`??`)
```
Comment (MarcoGorelli) [inline doc/source/whatsnew/v2.0.0.rst:492]: `:issue:`??``

use the PR number?

File: pandas/tests/tools/test_to_datetime.py [hunk 7]
```diff
@@ -1489,6 +1489,29 @@ def test_to_datetime_fixed_offset(self):
...
```
Comment (MarcoGorelli) [inline pandas/tests/tools/test_to_datetime.py:1512]: can this be constructed explicitly with `Index` / `DatetimeIndex` as you do above?

**Review by MarcoGorelli — COMMENTED** [review]
couple of minor comments, other than that looks good

File: pandas/tests/tools/test_to_datetime.py (hunk 7 above)
Comment (jbrockmendel) [reply pandas/tests/tools/test_to_datetime.py:1512]: i find this a lot simpler

Comment (MarcoGorelli) [reply pandas/tests/tools/test_to_datetime.py:1512]: same, but isn't it risk to have both `res` and `expected` go down nearly exactly the same path?

Comment (jbrockmendel) [reply pandas/tests/tools/test_to_datetime.py:1512]: the point of this test is that they behave the same.  some other test can check that the integer case behaves "correctly"

Comment (MarcoGorelli) [reply pandas/tests/tools/test_to_datetime.py:1512]: ok sure

**Review by MarcoGorelli — APPROVED** [review]
LGTM, thanks @jbrockmendel

File: pandas/_libs/tslib.pyx [hunk 8]
```diff
@@ -314,8 +330,33 @@ def array_with_unit_to_datetime(
...
+                        # Analogous to GH#47266 for Timestamp
+                        if is_raise:
+                            raise ValueError(
+                                f"Conversion of non-round float with unit={unit} "
+                                "is ambiguous"
+                            )
+                        elif is_ignore:
+                            raise AssertionError
+                        iresult[i] = NPY_NAT
+                        continue
```
Comment (MarcoGorelli) [inline pandas/_libs/tslib.pyx:356]: sorry to block this, but is there a test that hits this?

I can see tests when `val` is a non-round float, but what about when it's a string containing a non-round float (e.g. `'1.5'`), which I believe is what would reach this branch?

**Review by MarcoGorelli — CHANGES_REQUESTED** [review]

**Review by MarcoGorelli — APPROVED** [review]
Looks good to me, thanks @jbrockmendel !

@mroeschke any thoughts?

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @jbrockmendel

**Pull request merged**